### ISSUE TITLE
aws-node stuck in crash recycle attempting to use an ENI it deleted

### ISSUE THREAD

**Issue by peterbroadhurst** [aws/amazon-vpc-cni-k8s#359]
One of our nodes encountered an issue where the `aws-node` CNI pod is constantly crash recycling, and the ipamd logs show:
```
2019-03-22T16:12:20Z [ERROR] Failed to setup eni eni-0c18e47af27178549 network: failed to retrieve eni eni-0c18e47af27178549 ip addresses: fail to find eni addresses for eni eni-0c18e47af27178549: failed to describe network interface: InvalidNetworkInterfaceID.NotFound: The networkInterface ID 'eni-0c18e47af27178549' does not exist
```
Node is tainted to isolate while we investigate.

> We have a ticket open with AWS support, as the conclusion is the metadata API is returning stale data, and we have an ENI stuck in attaching status on this instance

I am raising this and will report root cause back once determined, in case others see the same symptoms.

Reviewing the earlier ipamd logs I can see that the ENI was released and deleted, but just a minute later the `Reconciling ENI/IP pool info` logic re-added the deleted ENI back into the pool

```
2019-03-22T14:26:53Z [DEBUG] FreeENI: found a deletable ENI eni-0c18e47af27178549
2019-03-22T14:26:53Z [INFO] FreeENI eni-0c18e47af27178549: IP address pool stats: free 14 addresses, total: 28, assigned: 14
2019-03-22T14:26:53Z [DEBUG] Start freeing eni eni-0c18e47af27178549
2019-03-22T14:26:53Z [INFO] Trying to free eni: eni-0c18e47af27178549
2019-03-22T14:26:53Z [INFO]  Pods deleted on my node: a0ptn8lpi5/a0ptn8lpi5-a0ox4r7pd8-quorum-node-0
2019-03-22T14:26:54Z [DEBUG] Found eni eni-0c18e47af27178549 attachment id: eni-attach-0263261d435bf0466
2019-03-22T14:26:54Z [DEBUG] Trying to delete eni: eni-0c18e47af27178549
2019-03-22T14:26:54Z [DEBUG] Not able to delete eni yet (attempt 1/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 17b480ad-6a54-4959-90da-ff6e9d7d443b
2019-03-22T14:26:59Z [DEBUG] Not able to delete eni yet (attempt 2/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: daf14218-0bf1-4b96-89da-429ce20bc141
2019-03-22T14:27:04Z [DEBUG] Not able to delete eni yet (attempt 3/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 3ae19a13-b6a2-4fad-9e55-ce8c523e1e8b
2019-03-22T14:27:10Z [DEBUG] Not able to delete eni yet (attempt 4/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 620d50cb-cd8f-4ed3-94b6-4267561b6eed
2019-03-22T14:27:15Z [DEBUG] Not able to delete eni yet (attempt 5/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 42968956-c4ef-4502-9b22-c2503c21086f
2019-03-22T14:27:20Z [DEBUG] Not able to delete eni yet (attempt 6/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 2b44be00-f4ab-40cf-a708-27eda1c16844
2019-03-22T14:27:25Z [DEBUG] Not able to delete eni yet (attempt 7/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 715129d9-f18f-4e4f-a3fb-a3d29c5c0a3e
2019-03-22T14:27:30Z [DEBUG] Not able to delete eni yet (attempt 8/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 6806f0a1-f7ac-4a27-8f60-8ddd6e9972b8
2019-03-22T14:27:35Z [DEBUG] Not able to delete eni yet (attempt 9/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 2bbf405b-5a90-4d37-9769-5f302dca5ee1
2019-03-22T14:27:40Z [DEBUG] Not able to delete eni yet (attempt 10/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 49b90fb4-110f-46e3-8352-8df5025ef64f
2019-03-22T14:27:46Z [DEBUG] Not able to delete eni yet (attempt 11/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: eeaf6d9f-4b71-4c27-bb7d-b0c7233fa4ce
2019-03-22T14:27:51Z [DEBUG] Not able to delete eni yet (attempt 12/20): InvalidParameterValue: Network interface 'eni-0c18e47af27178549' is currently in use.
        status code: 400, request id: 8cfad9c0-585c-498e-b3a1-fb2553b5eb21
2019-03-22T14:27:56Z [INFO] Successfully deleted eni: eni-0c18e47af27178549
2019-03-22T14:27:56Z [INFO] Successfully freed eni: eni-0c18e47af27178549
2019-03-22T14:27:56Z [DEBUG] Successfully decreased IP Pool
...
2019-03-22T14:28:56Z [DEBUG] Reconciling ENI/IP pool info...
2019-03-22T14:28:56Z [DEBUG] Total number of interfaces found: 3
2019-03-22T14:28:56Z [DEBUG] Found eni mac address : 06:42:d7:17:47:fe
2019-03-22T14:28:56Z [DEBUG] Using device number 0 for primary eni: eni-0c4084e42a1c0139c
2019-03-22T14:28:56Z [DEBUG] Found eni: eni-0c4084e42a1c0139c, mac 06:42:d7:17:47:fe, device 0
2019-03-22T14:28:56Z [DEBUG] Found cidr 10.12.64.0/20 for eni 06:42:d7:17:47:fe
2019-03-22T14:28:56Z [DEBUG] Found ip addresses [10.12.65.34 10.12.77.238 10.12.79.14 10.12.67.111 10.12.72.229 10.12.71.198 10.12.65.224 10.12.65.96 10.12.68.225 10.12.70.131 10.12.78.122 10.12.75.53 10.12.67.177 10.12.67.241 10.12.69.179] on eni 06:42:d7:17:47:fe
2019-03-22T14:28:56Z [DEBUG] Found eni mac address : 06:b1:c6:3f:cd:0e
2019-03-22T14:28:56Z [DEBUG] Found eni: eni-0c18e47af27178549, mac 06:b1:c6:3f:cd:0e, device 2
2019-03-22T14:28:56Z [DEBUG] Found cidr 10.12.64.0/20 for eni 06:b1:c6:3f:cd:0e
2019-03-22T14:28:56Z [DEBUG] Found ip addresses [10.12.66.86 10.12.68.168 10.12.72.201 10.12.77.234 10.12.71.139 10.12.74.165 10.12.70.166 10.12.77.31 10.12.75.121 10.12.67.154 10.12.78.27 10.12.74.244 10.12.79.85 10.12.76.54 10.12.79.145] on eni 06:b1:c6:3f:cd:0e
```

The `GetAttachedENIs()` function is calling the metadata API 

`http://169.254.169.254/latest/meta-data/network/interfaces/macs/`

to list the MACs of the attached ENIs, and then describing the device-number

`http://169.254.169.254/latest/meta-data/network/interfaces/macs/06:b1:c6:3f:cd:0e/device-number/`

and determining the ENI ID with

`http://169.254.169.254/latest/meta-data/network/interfaces/macs//06:b1:c6:3f:cd:0e/interface-id/`

The metadata service is **still** showing that MAC address 12 hours after it was deleted.

**Comment by mogren** [comment]
Thanks a lot @peterbroadhurst , that is a great find!

**Comment by mogren** [comment]
@peterbroadhurst Thanks, I've opened #458 with a change to stop using force, and just retry instead.

**Comment by mogren** [comment]
Merged fix to remove the force detach in #458. Will be in the v1.5.0 release.
